pkg/tasks/encryption_providers.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2019 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tasks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	kubeoneapi "k8c.io/kubeone/pkg/apis/kubeone"
+	"k8c.io/kubeone/pkg/scripts"
+	"k8c.io/kubeone/pkg/ssh"
+	"k8c.io/kubeone/pkg/state"
+	"k8c.io/kubeone/pkg/templates"
+
+	encryptionproviders "k8c.io/kubeone/pkg/templates/encryption-providers"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+	dynclient "sigs.k8s.io/controller-runtime/pkg/client"
+	kyaml "sigs.k8s.io/yaml"
+)
+
+// download the configuration from leader
+func FetchEncryptionProvidersFile(s *state.State) error {
+	s.Logger.Infof("Downloading EncryptionProviders configuration file...")
+	host, err := s.Cluster.Leader()
+	if err != nil {
+		return err
+	}
+
+	conn, err := s.Connector.Connect(host)
+	if err != nil {
+		return err
+	}
+	fileName := s.GetEncryptionProviderConfigName()
+	config, _, _, err := conn.Exec(fmt.Sprintf("sudo cat /etc/kubernetes/encryption-providers/%s", fileName))
+	if err != nil {
+		return err
+	}
+
+	s.LiveCluster.EncryptionConfiguration.Config = &apiserverconfigv1.EncryptionConfiguration{}
+	err = kyaml.UnmarshalStrict([]byte(config), s.LiveCluster.EncryptionConfiguration.Config)
+	return err
+}
+
+func UploadIdentityFirstEncryptionConficguration(s *state.State) error {
+	s.Logger.Infof("Uploading EncryptionProviders configuration file...")
+
+	if s.LiveCluster.EncryptionConfiguration == nil ||
+		s.LiveCluster.EncryptionConfiguration.Config == nil {
+		return errors.New("failed to read live cluster encryption providers configuration")
```

Comment:
Can we improve the error message here? It might sound something like:

```suggestion
		return errors.New("encryption providers configuration is already disabled")
```

Also, do we actually want to fail in such cases? If I understand correctly, this function is used only when disabling encryption providers, so if it's already disabled, then we don't really need to do anything unless I'm missing something.

Reply:
Not really. This is a defensive check. We should never be this far in the execution if the feature is disabled. If this check is true, something seriously wrong happened in the previous steps and we should fail.